File: harness/auto_receipts.py
"""Calm full-auto operator receipt copy (CLI / tests).

Mirrors the webapp StatusBar/CostBreakdown quiet language: budget meters and
halt/block wording that never imply a command ran or a compaction occurred
unless the halt reason itself says the objective was met.
"""
from __future__ import annotations

from typing import Any, Mapping, Optional


def _fmt_tokens(num: Any) -> str:
    try:
        n = float(num)
    except (TypeError, ValueError):
        return "0"
    if n < 0 or n != n:  # NaN
        return "0"
    if n >= 1_000_000:
        s = f"{n / 1_000_000:.1f}".rstrip("0").rstrip(".")
        return f"{s}M"
    if n >= 1000:
        s = f"{n / 1000:.1f}".rstrip("0").rstrip(".")
        return f"{s}k"
    return str(int(round(n)))


def format_budget_meters(snapshot: Optional[Mapping[str, Any]]) -> str:
    """Compact AutoBudget meters: ``2/20 swarms · 4.1k/50k tok · 45s``."""
    if not isinstance(snapshot, Mapping):
        return ""
    parts = []
    swarms_used = snapshot.get("swarms_used")
    max_swarms = snapshot.get("max_swarms")
    if isinstance(swarms_used, (int, float)) and isinstance(max_swarms, (int, float)):
        parts.append(f"{int(swarms_used)}/{int(max_swarms)} swarms")
    tokens_used = snapshot.get("tokens_used")
    if isinstance(tokens_used, (int, float)):
        max_tokens = snapshot.get("max_tokens")
        used = _fmt_tokens(tokens_used)
        if isinstance(max_tokens, (int, float)):
            parts.append(f"{used}/{_fmt_tokens(max_tokens)} tok")
        else:
            parts.append(f"{used} tok")
    elapsed = snapshot.get("elapsed_s")
    if isinstance(elapsed, (int, float)):
        parts.append(f"{max(0, int(elapsed))}s")
    return " · ".join(parts)


def format_auto_status_receipt(
    cycle: Any,
    snapshot: Optional[Mapping[str, Any]] = None,
) -> str:
    try:
        n = int(cycle)
    except (TypeError, ValueError):
        n = 0
    meters = format_budget_meters(snapshot)
    label = f"Full-auto · cycle {max(0, n)}"
    return f"{label} · {meters}" if meters else label


def format_auto_halt_receipt(
    reason: str,
    snapshot: Optional[Mapping[str, Any]] = None,
) -> str:
    raw = (reason or "").strip() or "Budget or policy ended the run"
    low = raw.lower()
    if "objective met" in low:
        label = "Full-auto finished"
    elif "cancel" in low:
        label = "Full-auto cancelled"
    elif any(k in low for k in (
        "ceiling", "stall", "killswitch", "token", "swarm", "idle", "seconds",
    )):
        label = "Full-auto halted"
    else:
        label = "Full-auto stopped"
    meters = format_budget_meters(snapshot)
    body = f"{label}: {raw}"
    return f"{body} · {meters}" if meters else body


def format_command_blocked_receipt(
    reason: str = "",
    category: str = "",
) -> str:
    detail = (reason or "").strip() or (category or "").strip() or (
        "Full-auto safety policy blocked this command"
    )
    return f"Command not run: {detail}"

File: webapp/src/lib/autoReceipts.ts
/**
 * Calm full-auto operator receipts — StatusBar / CostBreakdown quiet copy.
 *
 * These helpers never imply a command ran, a compaction occurred, or an
 * objective succeeded unless the halt reason itself says so.
 */

export type AutoBudgetSnapshot = {
  tokens_used?: number;
  max_tokens?: number;
  swarms_used?: number;
  max_swarms?: number;
  elapsed_s?: number;
  max_seconds?: number;
  idle_steps?: number;
  max_idle_steps?: number;
  halted?: string | null;
};

function formatTokens(num: number): string {
  if (!Number.isFinite(num) || num < 0) return "0";
  if (num >= 1_000_000) return (num / 1_000_000).toFixed(1).replace(/\.0$/, "") + "M";
  if (num >= 1000) return (num / 1000).toFixed(1).replace(/\.0$/, "") + "k";
  return String(Math.round(num));
}

/** Compact AutoBudget meters for a quiet chip (e.g. "2/20 swarms · 4.1k/50k tok · 45s"). */
export function formatAutoBudgetMeters(
  snapshot?: AutoBudgetSnapshot | null,
): string {
  if (!snapshot || typeof snapshot !== "object") return "";
  const parts: string[] = [];
  if (
    typeof snapshot.swarms_used === "number"
    && Number.isFinite(snapshot.swarms_used)
    && typeof snapshot.max_swarms === "number"
    && Number.isFinite(snapshot.max_swarms)
  ) {
    parts.push(`${snapshot.swarms_used}/${snapshot.max_swarms} swarms`);
  }
  if (typeof snapshot.tokens_used === "number" && Number.isFinite(snapshot.tokens_used)) {
    const used = formatTokens(snapshot.tokens_used);
    const max =
      typeof snapshot.max_tokens === "number" && Number.isFinite(snapshot.max_tokens)
        ? `/${formatTokens(snapshot.max_tokens)}`
        : "";
    parts.push(`${used}${max} tok`);
  }
  if (typeof snapshot.elapsed_s === "number" && Number.isFinite(snapshot.elapsed_s)) {
    parts.push(`${Math.max(0, Math.round(snapshot.elapsed_s))}s`);
  }
  return parts.join(" · ");
}

export function autoStatusPresentation(
  cycle: number,
  snapshot?: AutoBudgetSnapshot | null,
): { label: string; detail: string } {
  const n = Number.isFinite(cycle) ? Math.max(0, Math.round(cycle)) : 0;
  const meters = formatAutoBudgetMeters(snapshot);
  return {
    label: `Full-auto · cycle ${n}`,
    // Budget progress only — never "done", "compacted", or "executed".
    detail: meters,
  };
}

export function autoHaltPresentation(
  reason: string,
  snapshot?: AutoBudgetSnapshot | null,
): { label: string; detail: string; metObjective: boolean } {
  const raw = String(reason || "").trim();
  const metObjective = /objective met/i.test(raw);
  const cancelled = /\bcancel(?:led|ed)?\b/i.test(raw);
  const budgetTripped = /ceiling|stall|killswitch|token|swarm|idle|seconds/i.test(raw);

  let label = "Full-auto stopped";
  if (metObjective) label = "Full-auto finished";
  else if (cancelled) label = "Full-auto cancelled";
  else if (budgetTripped) label = "Full-auto halted";

  const detail = raw || "Budget or policy ended the run";
  const meters = formatAutoBudgetMeters(snapshot);
  return {
    label,
    detail: meters ? `${detail} · ${meters}` : detail,
    metObjective,
  };
}

export function commandBlockedPresentation(d: {
  reason?: string;
  category?: string;
}): { label: string; detail: string } {
  return {
    label: "Command not run",
    detail:
      (d.reason || "").trim()
      || (d.category || "").trim()
      || "Full-auto safety policy blocked this command",
  };
}

/** Truthful post-decision copy — never claims the shell command executed. */
export function commandApprovalStatusCopy(
  status: "pending" | "approving" | "approved" | "rejected" | "error",
): string {
  switch (status) {
    case "approving":
      return "Applying decision…";
    case "approved":
      return "Approved once — not run yet; retry queued.";
    case "rejected":
      return "Rejected — command was not run.";
    case "error":
      return "Decision failed — command remains blocked.";
    default:
      return "";
  }
}
